registry/remote/repository.go

```diff
@@ -1051,6 +1090,113 @@ func (s *manifestStore) push(ctx context.Context, expected ocispec.Descriptor, c
 	return verifyContentDigest(resp, expected.Digest)
 }
 
+// indexReferrersForPush indexes referrers for image or artifact manifest with the
+// subject field.
+// Reference: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-manifests-with-subject
+func (s *manifestStore) indexReferrersForPush(ctx context.Context, desc ocispec.Descriptor, r io.Reader) error {
+	var subject ocispec.Descriptor
+	switch desc.MediaType {
+	case ocispec.MediaTypeArtifactManifest:
+		var artifact ocispec.Artifact
+		if err := decodeJSON(r, desc, &artifact); err != nil {
+			return fmt.Errorf("failed to decode manifest: %s: %s: %w", desc.Digest, desc.MediaType, err)
+		}
+		if artifact.Subject == nil {
+			// no subject, no indexing needed
+			return nil
+		}
+		subject = *artifact.Subject
+		desc.ArtifactType = artifact.ArtifactType
+		desc.Annotations = artifact.Annotations
+	case ocispec.MediaTypeImageManifest:
+		var manifest ocispec.Manifest
+		if err := decodeJSON(r, desc, &manifest); err != nil {
+			return fmt.Errorf("failed to decode manifest: %s: %s: %w", desc.Digest, desc.MediaType, err)
+		}
+		if manifest.Subject == nil {
+			// no subject, no indexing needed
+			return nil
+		}
+		subject = *manifest.Subject
+		desc.ArtifactType = manifest.Config.MediaType
+		desc.Annotations = manifest.Annotations
+	default:
+		return nil
+	}
+
+	yes, err := s.repo.isReferrersAPIAvailable(ctx, subject)
+	if err != nil {
+		return err
+	}
+	if yes {
+		// referrers API is available, no client-side indexing needed
+		return nil
+	}
+
+	// there can be multiple go-routines updating the referrers tag concurrently
+	referrersTag := buildReferrersTag(subject)
+	s.repo.lockReferrersTag(referrersTag)
+	defer s.repo.unlockReferrersTag(referrersTag)
+
+	var updatedReferrers []ocispec.Descriptor
+	err = s.repo.referrersByTagSchema(ctx, subject, "", func(referrers []ocispec.Descriptor) error {
+		for _, r := range referrers {
+			// skip duplicate entry
+			if !content.Equal(r, desc) {
+				updatedReferrers = append(updatedReferrers, r)
+			}
+		}
+		return nil
+	})
+	if err != nil && !errors.Is(err, errdef.ErrNotFound) {
+		return err
+	}
+
+	updatedReferrers = append(updatedReferrers, desc)
+	indexDesc, index, err := generateReferrersIndex(updatedReferrers)
+	if err != nil {
+		return fmt.Errorf("failed to generate referrers index for referrers tag %s: %w", referrersTag, err)
+	}
+	return s.push(ctx, indexDesc, bytes.NewReader(index), referrersTag)
```

Comment:
After pushing, we need to delete previous index.

Reply:
Right, otherwise there will be many junk manifests.